### PR TITLE
store SparkCloud.devices in an expirable cache

### PR DESCRIPTION
**Pull request by spapin**
At SparkCloud initialisation, the call to the api that list devices is stored in class parameter.
If the spark disconnects, the parameter will still hold the original status.

This PR adds a caching on the spark device, this allow to detect a disconnect spark and to avoid calling a function or a parameter on it (which would yield a timeout)

### CODE REVIEW

File: spyrk/spark_cloud.py [hunk 1]
```diff
@@ -61,7 +62,7 @@ def __init__(self, username_or_access_token, password=None):
         seconds (per device?) to reply as it waits for an answer from the
         disconnected devices.
         """
-        self.spark_api = Hammock('https://api.spark.io')
+        self.spark_api = Hammock('https://api.particle.io')
```
Comment (spapin) [inline spyrk/spark_cloud.py]: Ah sorry about this, I really thought I cherry picked in another branch, this is not related to the expiring cache

**Comment by AxelVoitier** [comment]
Hello,

Thanks for this PR :).
I haven't had time yet to just try it.

But in the meantime I was wondering if a ttl of 10s was not too low?
The point of caching it in the first place was to avoid the delay of contacting the cloud everytime we access this property. Now that would not be everytime but every 5s. (It seems cached_property only reset the counter anytime the ttl elapsed).

I actually wonder how often this property is accessed in practice.
Some could always reuse it whenever accessing a device function/signal. And some could have the good idea of locally caching the accessor to a device function/signal, and so only looking at the devices property a few times at the beginning.


Cheers,
Axel

**Comment by spapin** [comment]
Cached_property will expire the device after 10 seconds, that means if a function is accessed less than once every 10 seconds, the device information will have to be reloaded. that means two calls to the spark api instead of one.

The spark api refresh rate for devices is 30 seconds (cant find reference for that anymore), I think 10 second timeout is a good compromise between noticing rapidly the spark disconnection and the impact on services with a high request rate.

Then you are right: if a user decide to store a Device object, then its status will never be refreshed and this PR would have no effect. However, with the current state, it is not possible to have a refreshed device once the SparkCloud instance has been created. I don't think it is a good behavior.

It should be fine to allow user to store locally the device instance or to call the device attribute every time ( and therefore refreshing it every 10 seconds) from the SparkCloud object. Maybe in this case using a cached_property is over engineering the problem, what do you think?

**Comment by spapin** [comment]
Hello,
What is the status on this PR?

**Comment by AxelVoitier** [comment]
Merged with latest changes.

Thank you, and sorry for the veryyy long delay ;(.
